File: mmoTest/Assets/Script/Data/Proto/GameLevel/Send/GameLevel_FailProto.cs
//===================================================
//作    者：边涯  http://www.u3dol.com  QQ群：87481002
//创建时间：2018-02-25 22:40:38
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 客户端发送战斗失败消息
/// </summary>
public struct GameLevel_FailProto : IProto
{
    public ushort ProtoCode { get { return 12005; } }

    public int GameLevelId; //游戏关卡Id
    public byte Grade; //难度等级

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(GameLevelId);
            ms.WriteByte(Grade);
            return ms.ToArray();
        }
    }

    public static GameLevel_FailProto GetProto(byte[] buffer)
    {
        GameLevel_FailProto proto = new GameLevel_FailProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.GameLevelId = ms.ReadInt();
            proto.Grade = (byte)ms.ReadByte();
        }
        return proto;
    }
}

File: MMORPGServer/Proto/CreateLuaProto/GameLevel_FailProto.lua
--客户端发送战斗失败消息
GameLevel_FailProto = { ProtoCode = 12005, GameLevelId = 0, Grade = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
GameLevel_FailProto.__index = GameLevel_FailProto;

function GameLevel_FailProto.New()
    local self = { }; --初始化self
    setmetatable(self, GameLevel_FailProto); --将self的元表设定为Class
    return self;
end


--发送协议
function GameLevel_FailProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.GameLevelId);
    ms:WriteByte(proto.Grade);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function GameLevel_FailProto.GetProto(buffer)

    local proto = GameLevel_FailProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.GameLevelId = ms:ReadInt();
    proto.Grade = ms:ReadByte();

    ms:Dispose();
    return proto;
end